### PR TITLE
Add frontend flag for explicitly setting ccc

### PR DESCRIPTION
**Pull request by rauhul**
Adds a new swift-frontend flag to allow users to choose which calling convention is used to make c function calls. This hidden flag is called `-experimental-platform-c-calling-convention`.

This behavior is needed to workaround rdar://109431863 (Swift-frontend produces trapping llvm ir for non-trapping sil). The root cause of this issue is that IRGen always emits c function calls with llvm's default C calling convention. However clang may select a different (incompatible) calling convention for the function, eventually resulting--via InstCombine and SimplifyCFG--in a trap instead of the function call. This failure mode is most readily seen with the triple `armv7em-apple-none-macho` when attempting to call functions taking struct arguments. Example unoptimized ir below:

```llvm-ir
call void @bar([4 x i32] %17, i32 2), !dbg !109
...
define internal arm_aapcs_vfpcc void @bar(
  [4 x i32] %bar.coerce, i32 noundef %x)
```

In the future it would be better to use the clang importer or some other tool to determine the calling convention for each function instead of setting the calling convention frontend invocation wide.

Note: I don't know for sure whether or not clang should be explicitly annotating these functions with a calling convention instead of aliasing C to mean ARM_AAPCS_VFP for this particular combination of `-target`, `-mfloat-abi`, and `-mcpu`.

### CODE REVIEW

File: include/swift/AST/IRGenOptions.h [hunk 1]
```diff
@@ -477,6 +478,9 @@ class IRGenOptions {
   /// function instead of to trap instructions.
   std::string TrapFuncName = "";
 
+  /// The calling convention used to perform non-swift calls.
+  llvm::CallingConv::ID ExperimentalPlatformCCallingConvention;
```
Comment (aschwaighofer) [inline include/swift/AST/IRGenOptions.h:482]: Lets drop the word `Experimental`.

**Review by aschwaighofer — COMMENTED** [review]
You'll also want to change:

```
lib/IRGen/GenHeap.cpp:  metadata->setCallingConv(llvm::CallingConv::C);
lib/IRGen/IRGenModule.cpp:  C_CC = llvm::CallingConv::C;
```

**Comment by rauhul** [comment]
@swift-ci please test

**Comment by rauhul** [comment]
@swift-ci please smoke test

**Review by aschwaighofer — APPROVED** [review]
Lgtm

**Pull request merged**